coffea/nanoevents/methods/physlite.py

```diff
@@ -0,0 +1,178 @@
+    779635413: "TruthBottom",
+}
+
+
+def _element_link(target_collection, global_index, key):
+    global_index = awkward.where(key != 0, global_index, -1)
+    return target_collection._apply_global_index(global_index)
+
+
+def _element_link_multiple(events, obj, link_field, with_name=None):
+    key = obj[link_field].m_persKey
+    unique_keys = [
+        i
+        for i in numpy.unique(awkward.to_numpy(awkward.flatten(key, axis=None)))
+        if i != 0
+    ]
+
+    def where(unique_keys):
+        target_name = _hash_to_target_name[unique_keys[0]]
+        links = events[target_name]._apply_global_index(
+            obj[f"{link_field}__G__{target_name}"]
+        )
+        if len(unique_keys) == 1:
+            return links
+        return awkward.where(key == unique_keys[0], links, where(unique_keys[1:]))
+
+    out = where(unique_keys).mask[key != 0]
+    if with_name is not None:
+        out = awkward.with_parameter(out, "__record__", with_name)
+    return out
+
+
+@awkward.mixin_class(behavior)
+class Particle(vector.PtEtaPhiMLorentzVector, base.NanoCollection):
+    @property
+    def mass(self):
+        return self.m
+
+
+_set_repr_name("Particle")
+
+
+@awkward.mixin_class(behavior)
+class TrackParticle(vector.LorentzVector, base.NanoCollection):
+    "see https://gitlab.cern.ch/atlas/athena/-/blob/21.2/Event/xAOD/xAODTracking/Root/TrackParticle_v1.cxx#L82"
+
+    @property
+    def theta(self):
+        return self["theta"]
+
+    @property
+    def phi(self):
+        return self["phi"]
+
+    @property
+    def p(self):
+        return 1.0 / numpy.abs(self.qOverP)
+
+    @property
+    def x(self):
+        return self.p * numpy.sin(self.theta) * numpy.cos(self.phi)
+
+    @property
+    def y(self):
+        return self.p * numpy.sin(self.theta) * numpy.sin(self.phi)
+
+    @property
+    def z(self):
+        return self.p * numpy.cos(self.theta)
+
+    @property
+    def t(self):
+        return numpy.sqrt(139.570 ** 2 + self.x ** 2 + self.y ** 2 + self.z ** 2)
+
+
+_set_repr_name("TrackParticle")
+
+
+@awkward.mixin_class(behavior)
+class Muon(Particle):
+    @property
+    def trackParticle(self):
+        return _element_link(
+            self._events().CombinedMuonTrackParticles,
+            self[
+                "combinedTrackParticleLink.m_persIndex__G__CombinedMuonTrackParticles"
+            ],
+            self["combinedTrackParticleLink.m_persKey"],
+        )
+
+
+_set_repr_name("Muon")
+
+
+@awkward.mixin_class(behavior)
+class Electron(Particle):
+    @property
+    def trackParticles(self):
+        return _element_link(
+            self._events().GSFTrackParticles,
+            self.trackParticleLinks__G__GSFTrackParticles,
+            self.trackParticleLinks.m_persKey,
+        )
+
+    @property
+    def trackParticle(self):
+        trackParticles = self.trackParticles
+        return self.trackParticles[
+            tuple([slice(None) for i in range(trackParticles.ndim - 1)] + [0])
```

Comment:
That did not wok - awkward complains with `ValueError: ellipsis (...) can't be used on a data structure of different depths`. Should that work? Then i can open an issue for awkward array.

Actually i noticed it for lower number of dimensions e.g. when i do `events[2].Electrons.trackParticle`

Reply:
I would think that as long as `array.ndim` is well-defined, ellipsis is as well (provided no field indexers are used in the getitem). @jpivarski thoughts?

Reply:
That error message about ellipsis is talking about an issue with unions (so, unless it's not what I'm thinking, it's not a bug).

```python
>>> array = ak.Array([[1], [2], [3]])
>>> array[..., 0]
<Array [1, 2, 3] type='3 * int64'>
>>> array = ak.Array([[1], [2], [3], [[4], [4], [4], [4]]])
>>> array[..., 0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jpivarski/miniconda3/lib/python3.8/site-packages/awkward/highlevel.py", line 993, in __getitem__
    return ak._util.wrap(self.layout[where], self._behavior)
ValueError: ellipsis (...) can't be used on a data structure of different depths

(https://github.com/scikit-hep/awkward-1.0/blob/1.2.2/src/libawkward/Content.cpp#L1491)
```

Arrays always have an `ndim`, and this is the "purelist depth". A union like the above doesn't have an easily definable depth. I _had thought_ that I made this return -1 in cases like that, but apparently, it gives the depth of the UnionArray object:

```python
>>> array.ndim                    # high-level
0
>>> array.layout.purelist_depth   # low-level
0
```

That's not ideal: it doesn't give more information than if it had been a RecordArray at that level. But I don't know if this addresses your question.

Reply:
I looked a bit what's going on for this particular array. Seems the issue here is not that it is a union, but that one field has a higher dimension. There is `definingParametersCovarianceMatrix` which has one dimension more than the other fields (but is still zipped on the outermost level of offsets). So a small example for that situation would be

```python
import awkward as ak
array = ak.zip(
    {"a" : [[1], [2, 3]], "b" : [[[10, 10]], [[20, 20], [30, 30]]]},
    depth_limit=2
)
```
The resulting array actually has `ndim=2` - which is what i use in that code for `trackParticle`.
```pycon
>>> array.ndim
2
```
I want this:
```pycon
>>> array[:, 0].tolist()
[{'a': 1, 'b': [10, 10]}, {'a': 2, 'b': [20, 20]}]
```
The ellipsis don't work, presumably because of the one dimension more in the field `b`
```
>>> array[..., 0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nikolai/.local/lib/python3.9/site-packages/awkward/highlevel.py", line 993, in __getitem__
    return ak._util.wrap(self.layout[where], self._behavior)
ValueError: ellipsis (...) can't be used on a data structure of different depths

(https://github.com/scikit-hep/awkward-1.0/blob/1.3.0/src/libawkward/Content.cpp#L1491)
```